### ISSUE TITLE
Migrating from FreeDOS to EDR bricked the PC

### ISSUE THREAD

**Issue by mateuszviste** [SvarDOS/edrdos#93]
A working PC running the FreeDOS kernel. Pulled the latest SYS and KERNLEDR from pkgnet. Deleted the FD kernel (`DEL C:\KERNEL.SYS`) and installed both packages. Then did a `SYS /BOOTONLY C:`. SYS said "system transferred" so I performed a reboot. Ended up with this:

![20240815_175554](https://github.com/user-attachments/assets/9576cbbc-f820-4fa7-b6cf-2d3e1c355e5e)

Any clue what went wrong?

**Comment by boeckmann** [comment]
As this is the single-file kernel, the message is a little bit misleading, to say at least. The most probable explanation is that EDR can not determin its boot device. Which SYS version did you use when upgrading?

**Comment by boeckmann** [comment]
Its boot partition, in particular.

**Comment by boeckmann** [comment]
The thing is that boot sectors of older sys versions did provide the boot drive in a register unexpected by EDR.

**Comment by boeckmann** [comment]
Another question: If you boot from floppy, is C: accessible.

**Comment by boeckmann** [comment]
The EDR kernel refuses to login partitions that exceed the size reported by the BIOS.

**Comment by mateuszviste** [comment]
And documentation (readme) says "v3.6f - Nov 20, 2024"

**Comment by mateuszviste** [comment]
Step 1

![20240815_184423](https://github.com/user-attachments/assets/d92df691-6bfe-486b-bc56-2a8b0044b447)


Step 2

![20240815_184457](https://github.com/user-attachments/assets/8260074d-bd23-43e9-b10c-f5d783d4f70d)

**Comment by boeckmann** [comment]
Is step 1 also under EDR kernel, or only step 2? And does to C: exceed the size reported by BIOS? Some BIOS report up to 3 cylinders less than actually there, but FDISK in some configurations does not care and tries anyway. Leading to an extra cylinder exeeding the reported size. These partitions would be rejected by EDR.

**Comment by mateuszviste** [comment]
> Is step 1 also under EDR kernel, or only step 2?

Step 1 is under MSDOS (booted from a floppy), but I explicitly gave SYS the C: drive as source, and there is the EDR kernel (and SvarCOM).

> And does to C: exceed the size reported by BIOS? Some BIOS report up to 3 cylinders less than actually there, but FDISK in some configurations does not care and tries anyway. Leading to an extra cylinder exeeding the reported size. These partitions would be rejected by EDR.

Good question, I will check that tomorrow.

**Comment by boeckmann** [comment]
I can relax this check to see if that is the case, and allow a +1 cylinder.

**Comment by mateuszviste** [comment]
That being said, when such sanity checks fail it would be nice for the kernel to output something. Not necessarily a human message in plain english, but at least some diagnostic code, symbol or whatever that would provide a hint about what is going on.

**Comment by ecm-pushbx** [comment]
I tested the file from http://svardos.org/repo/?a=pull&p=kernledr and it works, at least when booting from A: in qemu, using lDebug's `boot protocol edrdos drkernel.sys` or `boot protocol freedos drkernel.sys`, after I included the file in the boot image renamed to drkernel.sys of course.

You could try to boot into lDebug (off a diskette) then use one of the two boot commands I listed. Ordinarily you'd have to use `q` to unload lDebug and chain to the kernel. If you use `g` instead the debugger will start running the kernel. The kernel will run an `int3` that returns control flow to the debugger command line with `cs:ip = 0070:43D4`, at a `jb` marked as "not jumping". You can enter `r f CY` to change the kernel's behaviour so it doesn't re-init int 1 and int 3 handlers. This is part of the debugger support I added a while back. After changing the Carry Flag you can run an `r` command to make sure that the jump notice now says "jumping".

After that you can use run commands (`g`, `t`, `p`, `tp`) to trace until you get to the bits that can emit the error message. May be easier with a trace listing `.tls` file, a null modem serial cable, `install serial` in lDebug, and connecting TracList to the debugger output. I don't think Bernd's action artefacts have trace listings though. So you'd have to try a similar build from me instead.

**Comment by ecm-pushbx** [comment]
https://hg.pushbx.org/ecm/edrdos/rev/72d6c752ed4e?revcount=120#l1.69 This line can cause the error.

https://pushbx.org/ecm/download/old/edrdos/20240731.zip This is the build you'd want to test, with `drbio.tls` and `bin/edrdos.sys` for the simplest / most SvarDOS-ish experience.

**Comment by ecm-pushbx** [comment]
Oh of course to boot `kernel.sys` off `hda1` when you booted lDebug off a diskette you should specify `boot protocol freedos hda1/kernel.sys`

**Comment by boeckmann** [comment]
> https://hg.pushbx.org/ecm/edrdos/rev/72d6c752ed4e?revcount=120#l1.69 This line can cause the error.

Yes @ecm-pushbx I agree, most probably because the end cylinder of the partition is beyond the BIOS reported disk size, as C: is also inaccessible if he boots the kernel from A:

Reason why the partition may exceed reported BIOS size: FDISK used for partition creation did extra cylinder checks and used them (earlier Free FDISK versions defaulted to this)

**Comment by boeckmann** [comment]
I made a test with a 386 (86box) and forced the extra cylinder check with recent FDISK. Confirms that EDR then rejects the partition:

<img width="598" alt="Bildschirmfoto 2024-08-15 um 21 10 59" src="https://github.com/user-attachments/assets/bb896408-cf7e-4bbe-9dd7-a25044f991e5">
<img width="718" alt="Bildschirmfoto 2024-08-15 um 21 11 49" src="https://github.com/user-attachments/assets/0ed69bf6-22d8-4503-a780-a2edab2183f3">
<img width="380" alt="Bildschirmfoto 2024-08-15 um 21 12 16" src="https://github.com/user-attachments/assets/6421cdf7-25c0-4d44-bbad-454e206fac8e">

**Comment by mateuszviste** [comment]
@ecm-pushbx thanks for the idebug tips, that's very impressive and - honestly - I hope I won't have to debug it this way as idebug is very intimidating :)
About the EDR kernel: it works of course, I am using it since a couple of days already under different VMs and on a modern-ish (1999) PC. Today I tried it on my 386 and there it won't boot. Surely there is something specific about this PC. Bernd's HDD geometry theory is very convincing, esp. since I am fairly sure this disk was partitioned years ago with FreeDOS FDISK. I will check the geometry tomorrow, as the room where this PC lives is being occupied at the moment.

**Comment by ecm-pushbx** [comment]
> @ecm-pushbx thanks for the idebug tips, that's very impressive and - honestly - I hope I won't have to debug it this way as idebug is very intimidating :)

It's actually lDebug with a small L ;P

If you do need to debug something I'm always happy to answer questions. Such as for instance how to install bootable lDebug: [Invoking the debugger in boot loaded mode](https://pushbx.org/ecm/doc/ldebug.htm#invoking-boot) and some info on using instsect in [How to build the instsect application](https://pushbx.org/ecm/doc/ldebug.htm#buildinginstsect)

**Comment by boeckmann** [comment]
Floppy image with filesystem bounds check commented out. This will boot, hopefully

[edrdos.img.zip](https://github.com/user-attachments/files/16630015/edrdos.img.zip)

**Comment by boeckmann** [comment]
I documented this in an early issue: https://github.com/SvarDOS/edrdos/issues/26

**Comment by boeckmann** [comment]
The checks at
https://github.com/SvarDOS/edrdos/blob/dffa75ce9d8b92b61dbf8e37846e433261dd73d4/drbio/disk.asm#L2695-L2702 
seems to be contra productive, as a few lines later the kernel tries to read te boot sector anyway. And this would fail if the partition is REALLY out of bounds, making the explicit bounds check more harmful than beneficial?
https://github.com/SvarDOS/edrdos/blob/dffa75ce9d8b92b61dbf8e37846e433261dd73d4/drbio/disk.asm#L2705-L2712

So I think a reasonable "fix" would be to simply remove the check at L2695+?

**Comment by ecm-pushbx** [comment]
I think you should also remove the failure in https://hg.pushbx.org/ecm/edrdos/rev/72d6c752ed4e?revcount=120#l1.69 for the single-file load. Even if no hdd partition is found *and* we booted off a unit >= 80h it is still better to boot up with no / a wrong boot drive and possibly prompt the user for the shell pathname than to end up at the error display.

**Comment by boeckmann** [comment]
> I think you should also remove the failure in https://hg.pushbx.org/ecm/edrdos/rev/72d6c752ed4e?revcount=120#l1.69 for the single-file load. Even if no hdd partition is found _and_ we booted off a unit >= 80h it is still better to boot up with no / a wrong boot drive and possibly prompt the user for the shell pathname than to end up at the error display.

Yes!

For now I commented out in the test build for Mateusz the conditionals at
https://github.com/SvarDOS/edrdos/blob/dffa75ce9d8b92b61dbf8e37846e433261dd73d4/drbio/disk.asm#L2928-L2934
checking the partition end against to disk end the BIOS returned. If it works I would replace these conditional by a test read of the last partition sector. This should succeed even if the BIOS reports a few cylinders short.

**Comment by boeckmann** [comment]
@mateuszviste you may also test the image provided by https://github.com/SvarDOS/edrdos/actions/runs/10411274828. In this image, I replaced the dependency on BIOS size info by a test read of the last sector. If it succeedes the partition is not refused to login.

**Comment by mateuszviste** [comment]
@boeckmann your theory proves to be right. It is indeed a 1-cyl mismatch between the BIOS and the MBR.

This is what is configured in the BIOS:

![20240816_091140](https://github.com/user-attachments/assets/454bde70-c7f7-4ed8-9b4b-ada1388eceea)

And this is what FDISK reports (latest version from SvarDOS packages):

![20240816_090145](https://github.com/user-attachments/assets/9ed73366-ece7-4e1e-8075-b05eeeb7cb38)

While in the BIOS I had configured the disk as a 993 cylinders model, the BIOS seems to return one cylinder less to the OS. Why does it do that? Maybe it's afraid the outer cylinder is not reliable enough? Or it's just a stupid off-by-1 bug in the BIOS. Weird.

Anyway, the actual partition in the MBR spans over 993 cylinders, hence seemingly one cylinder past what the BIOS reports. Either because it was partitioned with the Free-FDISK version that performs a +1 cyl test, or because it was partitioned on a different PC, with a more truthful BIOS.

Either way - what should EDR be doing? This PC boots fine under both FreeDOS and MSDOS. It also boots fine with both EDR kernels that you suggested: the one from your image (with check disabled) and the one from github (with an active test). I do not think it is the kernel's job to test the HDD. Even if this +1 cylinder test fails, the kernel should still boot. Maybe it won't be able to load COMMAND.COM or CONFIG.SYS, but anything is better than a halt. Sure, the user might get into filesystem issues later if the CHS geometry is misconfigured, but that's not the kernel's fault.

**Comment by mateuszviste** [comment]
I suspected the landing zone being set to cyl 0 might make the BIOS subtract it from the available number of cylinders, but changing LZone to 994 did not change anything (FDISK still reports 992 cylinders instead of the 993 configured in the CMOS setup).

**Comment by boeckmann** [comment]
@mateuszviste thanks for testing. I have to admit its quite pleasing if a theory turns out to be right :)

As you noted, current FDISK do not check for extra cylinders by default when working in CHS mode. Additional info: if LBA support is enabled, FDISK never checks for extra cylinders. LBA BIOS functions should always return the correct size.

The BIOS reporting a few cylinder (up to three are reported) less on INT13,08 is "by design" and not an error. Some early hard disks used this cylinders for hardware tests etc.

What shall we do about the situation? Make a test read to the last sector and throw a warning if it is not accessible? Then at least the user is informed.

**Comment by mateuszviste** [comment]
> What shall we do about the situation? Make a test read to the last sector and throw a warning if it is not accessible? Then at least the user is informed.

I think this is fine, as long as it does not block the boot process. Another way would be to do nothing (don't test anything, don't check BIOS geometry, blindly believe what the MBR says). Both options are good in my book. :-)

**Comment by mateuszviste** [comment]
Quick note about the cylinder mismatch roots. Under [INT 13/AH=08h](http://www.ctyme.com/intr/rb-0621.htm) RBIL says:

> The maximum cylinder number reported in CX is usually two less than the total cylinder count reported in the fixed disk parameter table (see INT 41h,INT 46h) because early hard disks used the last cylinder for testing purposes; however, on some Zenith machines, the maximum cylinder number reportedly is three less than the count in the fixed disk parameter table.. For BIOSes which reserve the last cylinder for testing purposes, the cylinder count is automatically decremented.

**Comment by boeckmann** [comment]
> I think this is fine, as long as it does not block the boot process. Another way would be to do nothing (don't test anything, don't check BIOS geometry, blindly believe what the MBR says). Both options are good in my book. :-)

Kernel prints a warning if it can not access the last sector. Interestingly under 86box, BIOS returns success even if the sector is beyond the disk end?!? Nevertheless, worst thing seems to be that the user does not get the warning. So I think its ok. Can I close this?

**Comment by mateuszviste** [comment]
Looks very good. :) Thank you!
